topology: add macro to support multiple fir iir eq filters

fredoh9 opened the pull request:
The patch shows a proposal how to propagate pipeline filter
coefficients data definition at topology top level. It is an
useful way to avoid to create different new pipeline macros for
different filter. Define filter coefficients data with macro
DEF_EQFIR_PRIV and DEF_EQFIR_COEF, which will be replaced
unique name by adding PIPELINE number. For IIR filter, equivalent macro
is DEF_EQIIR_PRIV and DEF_EQIIR_COEF.

Makefile define each pipeline processing HSEARPROC, HSMICRPOC, SPKPROC,
DMICPROC etc. Each variable should support default value to minimize
current topologies. When new filter blob is required, define
PIPELINE_FILTERx in makefile. This method provides more flexible way
to support variable number of filters in the pipeline.

The other clean up for topology macros include definions for unique
volume tokens and unique filter coefficient definitions.

Signed-off-by: Fred Oh <fred.oh@linux.intel.com>

Related issues: https://github.com/thesofproject/sof/issues/2535, https://github.com/thesofproject/sof/issues/2534

fredoh9 commented on tools/topology/sof/pipe-eq-fir-volume-capture.m4:
```diff
@@ -0,0 +1,131 @@
+define(DEF_EQFIR_COEF, concat(`eqfir_coef_', PIPELINE_ID))
+define(DEF_EQFIR_PRIV, concat(`eqfir_priv_', PIPELINE_ID))
+
+# Fred: define filter set
```
I was adding more comments here, That changes are missed. I will collect all the review comments and process them together.

singalsu commented on tools/topology/CMakeLists.txt:
```diff
@@ -92,9 +92,11 @@ set(TPLGS
-	"sof-cml-rt5682\;sof-icl-rt5682\;-DPLATFORM=icl"
+	"sof-cml-rt5682\;sof-cml-rt5682\;-DPLATFORM=cml\;-DCPROC=volume"
+	"sof-cml-rt5682\;sof-cml-eq-rt5682\;-DPLATFORM=cml\;-DCPROC=eq-fir-volume"
+	"sof-cml-rt5682\;sof-cml-eq-lowpass-rt5682\;-DPLATFORM=cml\;-DCPROC=eq-fir-volume\;-DFILTERSET=lowpass"
```
Good, the XPROC for passing the pipeline macro is useful. I see the name CPROC infers capture processing but we should be able to separate DMIC, speaker, headset mic, headset ear endpoints and possibly define processing for several of them simultaneously. Could they be DMICPROC, SPKPROC, HSMICPROC, HSEARPROC?

In this case value lowpass for FILTERSET should be e.g. eq_fir_coef_lowpass_7khz_40db_48khz to call out the applied processing and simplify the generic pipeline macro. I wonder if the name usage could be like HSMICFILT=eq_fir_coef_xxxx. It would add m4 code to topology top level to redirect the processing to correct pipeline. We don't know pipeline numbers at makefile level so this kind of abstraction seems necessary.

singalsu commented on tools/topology/CMakeLists.txt:
```diff
@@ -92,9 +92,11 @@ set(TPLGS
-	"sof-cml-rt5682\;sof-cml-rt5682\;-DPLATFORM=cml"
-	"sof-cml-rt5682\;sof-whl-rt5682\;-DPLATFORM=whl"
-	"sof-cml-rt5682\;sof-icl-rt5682\;-DPLATFORM=icl"
+	"sof-cml-rt5682\;sof-cml-rt5682\;-DPLATFORM=cml\;-DCPROC=volume"
```
Maybe the need to specify volume could be avoided and define/default it in macro top level if missing? If volume needs to  be eliminated from pipeline -DxxxPROC could be set to passthrough to select pipe-passthrough-playback or capture.

singalsu commented on tools/topology/m4/eq_fir_coef_flat.m4:
```diff
@@ -1,5 +1,5 @@
 # Flat FIR EQ 19-Dec-2018
-CONTROLBYTES_PRIV(EQFIR_priv,
+CONTROLBYTES_PRIV(DEF_EQFIR_PRIV,
```
Good, I'll change the default in filter blobs generator to this string so there will no more need to hand edit these.

singalsu commented on tools/topology/m4/eq_iir_coef_flat.m4:
```diff
@@ -1,5 +1,5 @@
 # Flat IIR EQ 18-Dec-2018
-CONTROLBYTES_PRIV(EQIIR_priv,
+CONTROLBYTES_PRIV(DEF_EQIIR_PRIV,
```
Good, I'll change this into filter generator too.

singalsu commented on tools/topology/sof-cml-rt5682.m4:
```diff
@@ -45,7 +45,7 @@ PIPELINE_PCM_ADD(sof/pipe-volume-playback.m4,
 # Low Latency capture pipeline 2 on PCM 0 using max 2 channels of s24le.
 # 1000us deadline on core 0 with priority 0
-PIPELINE_PCM_ADD(sof/pipe-volume-capture.m4,
+PIPELINE_PCM_ADD(sof/pipe-`CPROC'-capture.m4,
```
So in cases like this we should use the endpoint specific macro. In this case if could be HSMICPROC.

The passing of filter blob is the challenge. Since we here know that pipeline number is 2. We could set macro PIPE2_FILT1 from HSMICFILT1.

paulstelian97 replied:
I am actually asking if quoting `CPROC' here will still allow it to be expanded. Maybe use `'CPROC`' to still allow expansion?

fredoh9 replied:
@paulstelian97 current code is expanded well. Otherwise it will throw error that it could find the file sof/pipe-CPROC-capture.m4.

singalsu commented on tools/topology/sof/pipe-eq-fir-capture.m4:
```diff
@@ -17,6 +17,9 @@ include(`eq_fir.m4')
+define(DEF_EQFIR_PRIV, concat(`eqfir_priv_', PIPELINE_ID))
+
 # Use flat response
 include(`eq_fir_coef_flat.m4')
```
This reads always flat and it's OK for this PR. We can align all available pipelines later. I will need them for testbench tests and I can start doing it once this system gets deployed.

singalsu commented on tools/topology/sof/pipe-eq-fir-volume-capture.m4:
```diff
@@ -0,0 +1,131 @@
+define(DEF_EQFIR_PRIV, concat(`eqfir_priv_', PIPELINE_ID))
+
+# Fred: define filter set
+ifelse(FILTERSET, `flat',
```
So, here the magic should happen. The if statement should check for first instantiated processing component if there's define for PIPEx_FILT1 (x = PIPELINE_ID) and include the m4 blob file. If it's not defined use "flat".

Would this kind of system work and  be maintainable? There could be also other ways but this is what I had in mind and it should be possible.

singalsu commented:
@fredoh9 I added with PR #2667 new FIR and IIR EQ m4 blobs those can be used for testing and as examples. There's responses like loudness and bass boost for FIR and IIR those are easy to verify subjectively or by measurement if set up in topology.

fredoh9 commented:
I updated to address @singalsu comments. I changed title and put `[WIP]` as I added some examples which may not need to go to master. Or I may need to add some more real life examples to demonstrate this changes. I'm pretty sure this provide very flexible way to add more topology with different piplines.

fredoh9 commented:
@singalsu 
HSMICPROC, HSEARPROC, DMICPROC are defined in sof-cml-rt5682 pipes.

Default response is flat for iir and fir filters. It can be set by makefile, PIPELINE_FILTER1 or PIPELINE_FILTER2.

CMakeLists.txt changes for demonstration for usage of  this PR.
```
"sof-cml-rt5682\;sof-cml-eq-fir-rt5682\;-DPLATFORM=cml\;-DHSMICPROC=eq-fir-volume"
	"sof-cml-rt5682\;sof-cml-eq-fir-loud-rt5682\;-DPLATFORM=cml\;-DHSMICPROC=eq-fir-volume\;-DPIPELINE_FILTER1=eq_fir_coef_loudness.m4"
	"sof-cml-rt5682\;sof-cml-eq-iir-rt5682\;-DPLATFORM=cml\;-DHSEARPROC=eq-iir-volume"
```

singalsu commented on tools/topology/CMakeLists.txt:
```diff
@@ -96,6 +96,9 @@ set(TPLGS
 	"sof-cml-rt5682\;sof-cml-rt5682\;-DPLATFORM=cml"
+	"sof-cml-rt5682\;sof-cml-eq-fir-rt5682\;-DPLATFORM=cml\;-DHSMICPROC=eq-fir-volume"
+	"sof-cml-rt5682\;sof-cml-eq-fir-loud-rt5682\;-DPLATFORM=cml\;-DHSMICPROC=eq-fir-volume\;-DPIPELINE_FILTER1=eq_fir_coef_loudness.m4"
+	"sof-cml-rt5682\;sof-cml-eq-iir-rt5682\;-DPLATFORM=cml\;-DHSEARPROC=eq-iir-volume"
```
The xPROC and PIPELINE_FILTERx look nice in the makefile.

Seems like the PR with new filter blobs is merged so we can use them.

Just need to think what examples to make. I'd set up HSMICPROC in every topology with eq-iir-volume with 40 Hz 0 dB high-pass. The high-pass is needed to prevent HDA codec headset glitches. Since the EQ in the capture does permanent impact the demo responses work better for playback.

A demo topology could be made with additional loudness IIR for headset. Also the speaker could get the eq-volume (should be renamed to eq-iir-eq-fir-volume for consistency) playback pipeline with IIR setup with pass and FIR set with pass response. I upstreamed the pass-trough blobs those switch the EQs to bypass. It consumes even less resources than flat.

fredoh9 replied:
Thank you for your comments. PIPELINE_FILTER1 to xPROC_FILTERy changes and handling from middle layer need extra effort to address. I will create a new issue and define scope of the work. 

The other comments are very reasonable and minor work to do. I can address them today.

singalsu commented on tools/topology/sof-cml-rt5682.m4:
```diff
@@ -19,11 +19,17 @@ include(`platform/intel/'PLATFORM`.m4')
+# if DMICPROC is not defined, default pipepline is volume
+ifdef(`HSMICPROC', , `define(HSMICPROC, volume)')
+ifdef(`HSEARPROC', , `define(HSEARPROC, volume)')
+ifdef(`DMICPROC', , `define(DMICPROC, passthrough)')
```
Since this is direct attach DMIC the default should be eq-iir-volume and filter blob 40 Hz highpass for 48 kHz for dmic01 and 40 Hz highpass for 16 kHz for dmic16k. 

This reveals need for added DMIC16KPROC and DMIC16K_FILTER1 macros handling to get different filter blob for 16 kHz pipeline. The DMICPROC could be reused but for flexibility better to have them separated.

singalsu commented on tools/topology/CMakeLists.txt:
```diff
@@ -96,6 +96,9 @@ set(TPLGS
 	"sof-apl-src-pcm512x\;sof-apl-src-pcm512x"
 	"sof-cml-rt5682\;sof-cml-rt5682\;-DPLATFORM=cml"
+	"sof-cml-rt5682\;sof-cml-eq-fir-rt5682\;-DPLATFORM=cml\;-DHSMICPROC=eq-fir-volume"
+	"sof-cml-rt5682\;sof-cml-eq-fir-loud-rt5682\;-DPLATFORM=cml\;-DHSMICPROC=eq-fir-volume\;-DPIPELINE_FILTER1=eq_fir_coef_loudness.m4"
```
I think the long lines can be split for better readability.

fredoh9 replied:
Not sure how to make multilines for one entry. Do you have any suggestion?

singalsu replied:
No I don't, this is OK.

singalsu commented on tools/topology/sof/pipe-eq-fir-volume-capture.m4:
```diff
@@ -0,0 +1,129 @@
+# Passthrough with volume Pipeline and PCM
```
Even if this would not be used for demo topologies please keep this included in your PR. It will be useful later.

singalsu commented:
Good work, thanks! I proposed some changes for processing setup for enhanced topologies. Also need to add the upper endpoint specific layer for PIPELINE_FILTERx. Also I realized that we need to be able to differentiate the filter configs for 48 kHz and 16 kHz DMIC DAI. But you've done most of the work so those are fairly small further edits edits.

singalsu commented:
The CI checks output will be interesting. Please update your PR if there are issues.

singalsu commented on tools/topology/sof-cml-rt5682.m4:
```diff
@@ -19,15 +19,22 @@ include(`platform/intel/'PLATFORM`.m4')
+# if XPROC is not defined, define with default pipe
+ifdef(`HSMICPROC', , `define(HSMICPROC, volume)')
+ifdef(`HSEARPROC', , `define(HSEARPROC, volume)')
+ifdef(`DMICPROC', , `define(DMICPROC, eq-iir)')
```
Sorry, I missed this. This needs to be eq-iir-volume too.

fredoh9 commented:
> Thanks, this is OK for now. In the next step when PIPELINE_FILTERx is definable per end-point we can make more interesting topology variants.

I will create new issue to follow up this.

fredoh9 commented:
> > Thanks, this is OK for now. In the next step when PIPELINE_FILTERx is definable per end-point we can make more interesting topology variants.
> 
> I will create new issue to follow up this.

https://github.com/thesofproject/sof/issues/2707 is created

fredoh9 commented:
SOFCI TEST

lgirdwood commented:
CI known issues.
